Split RHCOS into layers

jlebon opened the pull request:
This enhancement describes improvements to the way RHEL CoreOS (RHCOS) is built so that it will better align with image mode for RHEL, all while also providing benefits on the OpenShift side. Currently, RHCOS is built as a single layer that includes both RHEL and OCP content. This enhancement proposes splitting it into three layers. Going from bottom to top:
1. the (RHEL-versioned) bootc layer (i.e. the base rhel-bootc image shared with image mode for RHEL)
2. the (RHEL-versioned) CoreOS layer (i.e. coreos-installer, ignition, afterburn, scripts, etc...)
3. the (OCP-versioned) node layer (i.e. kubelet, cri-o, etc...)

The terms "bootc layer", "CoreOS layer", and "node layer" will be used throughout this enhancement to refer to these.

The details of this enhancement focus on doing the first split: creating the node layer as distinct from the CoreOS layer (which will not yet be rebased on top of a bootc layer). The two changes involved which most affect OCP are:
1. bootimages will no longer contain OCP components (e.g. kubelet, cri-o, etc...)
2. the `rhel-coreos` payload image will be built in Prow/Konflux (as any other)

Tracked at: https://issues.redhat.com/browse/OCPSTRAT-1190

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+
+Diving into the first part then, at a high-level this is split into three phases.
+
+#### Phase 1: Using RHEL/CentOS Stream-only bootimages
```
Isn't there a "Phase 0" we need to link to here of shipping the result of https://github.com/openshift/os/pull/1445 in the release image? Is that already happening and I missed it?

```
$ oc adm release info registry.ci.openshift.org/ocp/release:4.17.0-0.nightly-2024-06-07-045541 |grep rhel
  rhel-coreos                                    sha256:7ab7b8bdb9f6ddde8d3f860324cacbb2e2ec7438b4228aadbf6fcb71dc1aa307
  rhel-coreos-extensions                         sha256:9af9706fa436bfc95940ef165c7a90302ce8907e5fcdb59e6cbdf542d7ba2d47
```

I expected to see something there like `rhel-coreos-base` or something that we can clearly say that the bootimages are instantiated from.

cgwalters replied:
Because in practice it *is* there, it's just not explicit.  And if the bootimages are deriving from it and not the rhel-coreos one I just think it'd be generally useful to be able to directly see it.

Mainly just for dev/support cases, there wouldn't be a customer facing case for consuming it.

jlebon replied:
Hmm, not sure that's worth adding to the release payload for. Is there precedence for shipping images in the release payload that are only useful for support cases?

jlebon replied:
Marking as resolved for now to clean up threads. Feel free to un-resolve and add more information.

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+
+#### Phase 1: Using RHEL/CentOS Stream-only bootimages
+
+In this phase, the goal is to have the OpenShift installer use bootimages containing only
```
xref e.g. https://github.com/openshift/installer/pull/6443 which changed the bootstrap node to rebase already for OKD, and we could then make that generic.

jlebon replied:
If you don't mind, let's resolve this thread in favour of the others already related to this.

jlebon commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+  - "@jlebon"
+  - "@cverna"
+  - "@travier"
+reviewers:
+  - "@patrickdillon, for installer impact"
+  - "@rphillips, for node impact"
+  - "@joepvd, for ART impact"
+  - "@sinnykumari, for MCO impact"
+  - "@LorbusChris, for OKD impact"
+  - "@zaneb, for agent installer impact"
+  - "@sdodson, for overall architecture"
+  - "@cgwalters, for overall architecture"
+approvers:
+  - "@mrunalp"
```
Apparently, the bot won't automatically tag the folks listed here, so manually doing it: @patrickdillon @rphillips @joepvd @sinnykumari @LorbusChris @zaneb @sdodson @cgwalters.

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+In this phase, the goal is to have the OpenShift installer use bootimages containing only
+RHEL/CentOS Stream content.
+
+To do this, we will start building two new streams in the RHCOS pipeline containing only pure
+RHEL/CentOS Stream content (let's call these the "pure RHEL stream" and "pure CentOS stream").
+Those streams will also be building the usual bootimages and uploading them to cloud.
```
Per above, IMO we need a clear, concise and understandable term for this...I don't think many people understand "stream" in this context (even though it is applicable!).  I was straw-manning `rhel-coreos-base` above.  And to extend on that, I think there's a bit of an open question here whether in the bigger picture the pipeline really needs to churn bootimages as often given that we are producing supported tooling for materializing on-demand disk images from container images.

So in the more medium term I'd advocate trying to create a clearer split between containers and disk images versus rolling them into a "stream" concept.

jlebon replied:
Yeah, wanted to avoid mentioning streams here, since it's so overloaded, but when it gets to implementation details, it'd be weird to not discuss it. I'll rework it to clarify that we're talking about "CoreOS pipeline streams" and not something else.

Re. lowering cadence of bootimages, yes that's definitely an overall goal of this. Part of it will happen naturally simply by the fact that we're decoupling OCP churn from the bootimages. The other part will be through conscious efforts (need to pick up https://github.com/coreos/fedora-coreos-pipeline/pull/810 again). But I'd say this is more CoreOS pipeline implementation details.

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+RHEL/CentOS Stream content (let's call these the "pure RHEL stream" and "pure CentOS stream").
+Those streams will also be building the usual bootimages and uploading them to cloud.
+
+Once we have these bootimages, we can better start adapting components that will need it
+to account for the lack of OpenShift components in the bootimages. Likely suspects here are
+any components involved in the bringup of the cluster (installer, Assisted Installer, MCO, etc...).
```
We have three choices broadly:

- Per above, pull and reboot into OCP container image, what is already done in installer today for OCP.  Note a subtle but very important point today...OCP sometimes ships kernels outside of RHEL cycles, and while that's not *usually* relevant for bootstrap, I wouldn't say it never would be
- Special case bootstrap to just e.g. `bootc usroverlay + dnf install`
- Run the installation process in a container (IMO...this is actually best)

travier replied:
That reads to me like a next step after this enhancement? Ideally we don't change the cluster installation flow (yet) and complete this transition first before looking at other installation options?

cgwalters replied:
> That reads to me like a next step after this enhancement? Ideally we don't change the cluster installation flow (yet) and complete this transition first before looking at other installation options?

If the bootimages are changed to not contain OCP content as the enhancement says, then things will definitely need to change in the install process.

Or, that bit could be moved to a later phase, which also seems viable to me (but I think also lowers the value of things, because being able to reuse the same bootimages for *multiple* OCP versions would be quite a big improvement).

jlebon replied:
> Note a subtle but very important point today...OCP sometimes ships kernels outside of RHEL cycles, and while that's not usually relevant for bootstrap, I wouldn't say it never would be

Can you expand on this? E.g. FIPS/crypto-related things? I could imagine detecting those cases where we should reboot and only do so then. Though in the common case it'd be very desirable to try to avoid rebooting.

> Run the installation process in a container (IMO...this is actually best)

Doesn't that imply trying to run the kubelet containerized? Not sure how supported that is nowadays.

> Special case bootstrap to just e.g. `bootc usroverlay + dnf install`

Yeah, definitely a potential path forward that's not too complex. We could ship the kubelet RPM in the extensions container image, which is already in the payload.

Hmm, or another approach is to just copy the kubelet out of `rhel-coreos`, since it's almost static anyway. There's still glibc, but I don't think mismatch issues would be a concern here (this isn't the "new worker node" flow with arbitrarily old bootimages, here we're doing an install so we should be guaranteed to be using matching images + it's a known combination that can be extensively tested in CI).

cgwalters replied:
> Doesn't that imply trying to run the kubelet containerized? Not sure how supported that is nowadays.

Yes, but the important thing here is that we don't need any non-core kubelet functionality.  In particular, CSI (storage) drivers running containerized I think were the #1 problem.  But we don't need any of that, it just needs to run pods.

There is also the angle that nowadays we could just pass pod definitions directly to podman...which I don't believe existed or was really fleshed out when OCP 4.0 was being designed.  That may even be simplest.

jlebon replied:
I added this as an open question, but keeping this thread unresolved since that's where all the context.

eranco74 replied:
> There is also the angle that nowadays we could just pass pod definitions directly to podman...which I don't believe existed or was really fleshed out when OCP 4.0 was being designed. That may even be simplest.

`podman play kube` definitely sounds like the simplest way to get the current bootstrap flow working without kubelet.

eranco74 replied:
> Hmm, or another approach is to just copy the kubelet out of rhel-coreos, since it's almost static anyway. There's still glibc, but I don't think mismatch issues would be a concern here

This might lead bugs such as: e.g. https://issues.redhat.com/browse/MGMT-16705 ( in this case the kubelet is incompatible because it's doing late-binding hence the node was booted with a default RHCOS and the not RHCOS of the release, yet the kubelet.conf is the one for the release).

zaneb replied:
> Likely suspects here are any components involved in the bringup of the cluster

I think it would be more helpful to focus on stages rather than components. There are 3 that will need to work. In reverse order:

1. Booting a host as a node after it is installed. Presumably this is a fairly straightforward matter of having it boot the OCP layer from the release payload.
2. Running the bootstrap node. This normally depends on OCP components. In the Assisted/ABI case, it necessarily runs on the base image, and we cannot stop services that were already running on the host prior to bootstrap starting.
3. The Assisted discovery ISO/ABI agent ISO. I believe this is a no-op, as neither of them should depend on OCP components today.

jlebon replied:
> > Hmm, or another approach is to just copy the kubelet out of rhel-coreos, since it's almost static anyway. There's still glibc, but I don't think mismatch issues would be a concern here
> 
> This might lead bugs such as: e.g. [issues.redhat.com/browse/MGMT-16705](https://issues.redhat.com/browse/MGMT-16705) ( in this case the kubelet is incompatible because it's doing late-binding hence the node was booted with a default RHCOS and the not RHCOS of the release, yet the kubelet.conf is the one for the release).

Is this something we should support though? Having the bootstrap process also have to worry about different bootimage versions instead of only the blessed, heavily CI-tested, version in `rhcos.json` doesn't sound great.

jlebon replied:
> Running the bootstrap node. This normally depends on OCP components. In the Assisted/ABI case, it necessarily runs on the base image, and we cannot stop services that were already running on the host prior to bootstrap starting.

What does ABI refer to here?

Can you expand/link to more information re. the "we cannot stop services that were already running on the host prior to bootstrap starting"? Don't we own the bootstrap node?

zaneb replied:
ABI = agent-based installer

In both assisted and ABI installs, the bootstrap-in-place runs inside the live ISO that is already running the agent. I believe assisted-service doesn't take well to the agent being restarted, but ABI has even tighter requirements: assisted-service itself is running on the bootstrap node, so interrupting it will break the entire installation process.

jlebon replied:
With https://github.com/openshift/installer/pull/8742 and https://github.com/openshift/assisted-installer/pull/899, I think we have a path forward now on this. Resolving.

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+
+#### Hypershift / Hosted Control Planes
+
+TODO
```
One thing I've mentioned in other places here is that if we can get to the point of shrinking our use case for Ignition to be very small by using container images instead, then we don't require the MCS anymore, which would definitely simplify Hypershift.

jlebon replied:
I'll add a note for that, though probably in the "Follow-ups" section instead.

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+
+#### Standalone Clusters
+
+TODO
```
This is what the enhancement is currently about right?  I think we can just state:

> This enhancement covers standalone clusters.

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+
+#### Single-node Deployments or MicroShift
+
+TODO
```
I think we can clearly say that the overarching changes here are going to *much* more strongly align MicroShift with OCP.

I also think (hope, pretty confidently) we aren't going to break anything with SNO here and will trend towards improving that too.

jlebon replied:
Yeah, I wonder if it'd make sense to share with MicroShift the node layer definition file (i.e. the MicroShift layers would be: bootc layer, node layer, MicroShift layer). Would be cool to have it actually derive from the OCP node image, but it definitely includes a lot of stuff not needed there so that doesn't make sense.

For SNO, the way [it does bootstrap-in-place](https://github.com/openshift/enhancements/blob/master/enhancements/installer/single-node-installation-bootstrap-in-place.md) using the live ISO I think should guide how we implement bootstrapping so that it works just as well there as it does in the standalone flow (which also then argues for not rebooting).

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+
+The CoreOS layer definition at this point would probably better live somewhere under gitlab.com/redhat
+alongside other rhel/centos-bootc projects. Then, openshift/os would be free to contain even more
+OpenShift-specific things. For example, we might consider moving content currently in the MCO's
+templates into openshift/os. or we might consider rewriting things currently written in bash
+into a statically-typed language that can literally live in openshift/os and get compiled and
+added as part of the derived (now multi-staged) layered build.
```
I think this is going to be really nice personally!

cgwalters commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+
+### Risks and Mitigations
+
+TODO
```
One thing I've learned here when we first did the layering switch is that it's a bit like changing the jet engine mid flight, and this is going to be bigger than that.  OTOH, we know more, and we have a plan.

jlebon replied:
Yeah, I have some items to add to this section!

cgwalters reviewed (commented):
Awesome work on this!

sdodson commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+- Share efforts, especially CI and docs, with the larger rhel/centos-bootc project
+- Maintain a single RHCOS stream per RHEL stream
+- Build the OpenShift node image as a layered image on top of RHCOS/SCOS
+- Build the OpenShift node image using Prow (in CI) and Konflux (in prod)
```
```suggestion
- Introduce no additional reboots in bootstrapping
```
I think reading through things that's the case, we'd just be going from a CoreOS layer without a Kubelet directly to a node layer with kubelet and all MachineConfig applied. Does that mean then that at least a stub MCD still lives in CoreOS Layer?

sdodson replied:
Or if there's no MCD in these images, can we make that more clear?

sdodson replied:
Executed by podman or something that's in the bootc layer or CoreOS layer in this enhancement?

jlebon replied:
Yes, podman is still going to be in the bootimages.

> Introduce no additional reboots in bootstrapping

I think we should be able to pull that off but it will require some design and testing. Worth highlighting [Colin's comment](https://github.com/openshift/enhancements/pull/1637/files#r1631394952) that OKD at least does this already. But for OCP, yes this would be a regression and would add time to the install flow (esp. bare metal).

mhrivnak replied:
Am I interpreting this discussion correctly that this change will require an additional reboot during install, and so you are not going to accept Scott's proposed goal?

I'm just confused by the statement "we should be able to pull that off" followed by "this would be a regression and would add time to the install flow". But I might just be missing some context.

sdodson replied:
Just noticing that my original suggestion replaced line 105 which was not my intention. I've updated it to just add line 106.

jlebon replied:
I've made it an explicit goal not.

sdodson commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+benefits on the OpenShift side. Currently, RHCOS is built as a single layer that
+includes both RHEL and OCP content. This enhancement proposes splitting it into
+three layers. Going from bottom to top:
+1. the (RHEL-versioned) bootc layer (i.e. the base rhel-bootc image shared with image mode for RHEL)
+2. the (RHEL-versioned) CoreOS layer (i.e. coreos-installer, ignition, afterburn, scripts, etc...)
+3. the (OCP-versioned) node layer (i.e. kubelet, cri-o, etc...)
```
```suggestion
1. bootc layer -- A RHEL-versioned layer (i.e. the base rhel-bootc image shared with image mode for RHEL)
2. CoreOS layer -- A RHEL-versioned layer (i.e. coreos-installer, ignition, afterburn, scripts, etc...)
3. node layer -- The OCP-versioned layer (i.e. kubelet, cri-o, etc...)
```
Just to move the terms used throughout to a bit more prominence.

jlebon commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+additional testing, this layered image will replace the current `rhel-coreos` image in the
+production release payload.
+
+### Workflow Description
```
Not sure about that. I think it would still make sense on https://mirror.openshift.com/ to have the bootimages accessible through OCP-versioned directories (but e.g. it'd be silly not to symlink them on the server side). I think that would reduce the likelihood that users pick the wrong bootimages.

jlebon replied:
@travier Any thoughts on this? Should we resolve this?

rphillips commented:
openshift/kubernetes has a specific workflow where jobs will build a new kubelet to use during the job run. This helps with rebase work and validating new kubernetes versions coming into OpenShift. We should preserve this workflow when migrating to RHCOS layering.

/cc @soltysh

jlebon commented:
> openshift/kubernetes has a specific workflow where jobs will build a new kubelet to use during the job run. This helps with rebase work and validating new kubernetes versions coming into OpenShift. We should preserve this workflow when migrating to RHCOS layering.
> 
> /cc @soltysh

I don't expect any issues there. That workflow should keep working as is.

zaneb commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+Another important follow-up would to be to hook up better CI testing to openshift/os. There is
+currently no CI test on that repo which actually launches a cluster. The reason is that RHCOS
+is just built differently. But now that the node image is simply yet another layered image build,
+it fits perfectly in Prow's opinionated model of building the image and shoving it in the test
```
Worth noting that the bootstrap node can't be tested this way. (The final node can, which is great.)

zaneb commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+  into the node image before continuing with bootstrapping, but that incurs an additional
+  reboot which is against the stated goals. Some other possibilities so far:
+  - Run the kubelet as a container
+  - Use podman to run the static pods
```
FWIW my experience with `podman kube play` a couple of years ago was a bit frustrating. There wasn't a way to generate the correct systemd service file automatically (`podman systemd generate` didn't work for kube-played pods), and having all containers in the pod running under a single systemd service was not great - systemd can only watch one conmon, so you lost the ability to granularly control what happens when a container within the pod dies. Logs for the different containers aren't separated in the journal.
Kube Play support in quadlet is a big step forward (not sure if that landed in RHEL 9 yet?), but if there is any internal complexity in the pods we are starting then we might still come to regret relying on it.

mhrivnak replied:
quadlet does work on RHEL 9.

zaneb replied:
Yeah, looks like Kube unit in quadlet [was there](https://docs.podman.io/en/v4.4/markdown/podman-systemd.unit.5.html#kube-units-kube) when quadlet first rolled out in podman 4.4.
I misremembered, thinking it was only added later. It was Pod unit that was added later, in podman 5.0.

rphillips replied:
There are specific shutdown requirements for static and regular pods. The kubelet also has to report status and mirror pods to the API. I do not believe having podman run static pods is a viable solution.

jlebon replied:
@rphillips And to be clear, we rely on these parts of the kubelet surface even just in the bootstrapping phase? And it would be too onerous to try to adapt to the gaps that switching to `podman kube play` would create?

rphillips replied:
Correct. Static pods run in a normal OpenShift cluster, mostly: kube-apiserver, controller-manager, and etcd. Networking may have 1-2 as well. All these pods are reported to the API server, and operators interact with configuring and managing them via the K8S API.

`podman kube play` might be able to run them, and could be written to follow the correct termination sequence; however, the Kubelet reports the static pods to the API as "mirror" pods. If we were to go down this path, then there would be a lot of re-writing core K8S functionality.

zaneb commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+- How should we adapt the bootstrapping process of the installer to handle the lack
+  of oc and kubelet in the bootimage? The easy approach would be to download and pivot
+  into the node image before continuing with bootstrapping, but that incurs an additional
+  reboot which is against the stated goals. Some other possibilities so far:
```
This is a non-starter for ABI and Assisted.

zaneb commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+  reboot which is against the stated goals. Some other possibilities so far:
+  - Run the kubelet as a container
+  - Use podman to run the static pods
+  - Install the kubelet RPM
```
From where?

rphillips replied:
This is probably the best approach. OpenShift would have to publish an RPM repo. Another issue that would arise are offline installs; where do they get oc and the kubelet.

zaneb replied:
I think the rpm repo would have to be running in a container from the release payload.
Even finding that without `oc` is challenging though, since it involves digging through the metadata to find the right image.

jlebon replied:
We actually already ship RPMs today in the extensions container (part of the payload). I'm not sure yet whether it makes sense to ship the kubelet there (the package set is 95% RHEL, so ideally it would also be RHEL-versioned), but it's not hard to add another container image similar to it.

I think this is probably the lowest friction approach. It's just awkward to have the kubelet RPM in a container image *and* in the node image. It's not very far at that point from just [copying the kubelet out of the image](https://github.com/openshift/enhancements/pull/1637#discussion_r1635218737).

> Even finding that without oc is challenging though, since it involves digging through the metadata to find the right image.

That part should be fine. `image_for` uses podman, and we can run `oc` using podman as well. Basically: `podman run ... $(image_for rhel-coreos) oc ...`.

rphillips replied:
For reference, it was just recently the hyperkube rpm [[spec](https://github.com/openshift/kubernetes/blob/master/openshift.spec#L72-L77)] was refactored to split controller-manager, api-server, and kubelet into separate RPMs. We used to ship api server and controller-manager in RHCOS, but never used them there. The container image for those components would run it out of the image. Since the recent change, the kubelet rpm is injected, but it wasn't always the case.

I worry about copying the kubelet out of the image due to openssl, fips, glibc , and perhaps other dependencies. I might vote 'awkward' is ok if it's just a duplication of a binary. The alternative of copying the binary out of an image seems a bit more brittle.

zaneb commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+  - Run the kubelet as a container
+  - Use podman to run the static pods
+  - Install the kubelet RPM
+  - `bootc apply-live`
```
Would this result in running systemd services being stopped? If so this is a non-starter for ABI and Assisted.

jlebon replied:
It wouldn't in this case, no. (Also I didn't mention it there, but this would be `rpm-ostree apply-live` for now; https://github.com/containers/bootc/issues/76).

But actually, I don't think this will work for the ISO case currently. Or at least, it'd need some work I think on the bootc/rpm-ostree side first.

zaneb commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+  - Use podman to run the static pods
+  - Install the kubelet RPM
+  - `bootc apply-live`
+  - `systemctl soft-reboot`
```
This is a non-starter for ABI and Assisted.

mrunalp replied:
Do we have more details for ABI/Assisted and an option which will work for those scenarios?

zaneb replied:
> Do we have more details for ABI/Assisted

How long have you got? :upside_down_face: 

> and an option which will work for those scenarios?

There's a discussion in the main thread that seems promising: https://github.com/openshift/enhancements/pull/1637#issuecomment-2240140391

zaneb commented:
/cc @cybertron @andfasano

soltysh commented:
> I don't expect any issues there. That workflow should keep working as is.

I believe this was the pre-req work done in https://github.com/openshift/kubernetes/pull/1805, which ensured we won't have problems in o/k.

rphillips commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+  of oc and kubelet in the bootimage? The easy approach would be to download and pivot
+  into the node image before continuing with bootstrapping, but that incurs an additional
+  reboot which is against the stated goals. Some other possibilities so far:
+  - Run the kubelet as a container
```
Upstream is not supporting running kubelet in a container anymore. CoreOS and other K8S distributions ran Kubelet in a container many many years ago. The lessons learned were not to do it.

mrunalp replied:
Yes, we can rule this one out.

romfreiman commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+
+* As a CoreOS engineer, I want to build RHCOS on top of rhel-bootc, so that I can benefit from support by the wider RHEL organization on those images and collaborate on CI testing and fixing bugs.
+
+* As an OpenShift customer, I want to be able to follow (most of) the image mode for RHEL documentation and use (most of) the same tooling surrounding it when customizing RHCOS, so that I have a consistent experience across image mode for RHEL and RHCOS and a clear understanding of how they relate to each other.
```
As an hardware vendor, I want to run RHEL certification process once and claim that OCP is certified as well.

romfreiman replied:
Seems to be mentioned as long term goal so we can resolve

romfreiman commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+  - "@LorbusChris, for OKD impact"
+  - "@zaneb, for agent installer impact"
+  - "@sdodson, for overall architecture"
+  - "@cgwalters, for overall architecture"
```
@fabiendupont @bthurber @ybettan please review in a context of kmm and dtk

bthurber replied:
Since KMM and DTK are day 2 in the case of OCP, I don't see any real impact or change here.  We aren't managing anything as day 0.

jlebon commented:
OK, so let's resume the bootstrapping issue. Restating some of the things from above and from researching further:
- We can't run the `kubelet` in a container because it's no longer supported.
- The delta between `kubelet` and `podman play` is too large to make the latter a feasible replacement.
- `systemctl soft-reboot` is not in RHEL9.
- In the AI/ABI/SNO cases, bootstrapping happens *in the live environment* where e.g. rebooting is not possible.
- I considered cobbling something around `kexec`, but in the limit, there are potential issues with kexec and hardware reliability, as well as how it meshes with Secure Boot.

What I'm playing with now is basically to have a special `node-image-pivot.target` that the node isolates to first. There, we pull the node image, unencapsulate it, check out its contents, and then mount over /usr and do a rough 3-way /etc merge. We then isolate back to multi-user.target to continue with the bootstrapping process.

This is in effect like a more aggressive `bootc/rpm-ostree apply-live`, though that [doesn't currently work in live environments](https://github.com/coreos/rpm-ostree/issues/4547). (Though even in the non-live case, there are some issues there that would need to be resolved.) It's close to [what OKD currently does when using FCOS live media today](https://github.com/openshift/installer/blob/b2876fc66fb8292443a3ff19d6db2d9d829e09f2/data/data/bootstrap/files/usr/local/bin/bootstrap-pivot.sh.template#L46), though using the ostree stack and isolating targets should make this more robust.

WIP for this in https://github.com/openshift/installer/pull/8742.

rphillips commented:
@jlebon That sounds like it might work. Where will the Kubelet be coming from? An OpenShift built image?

zaneb commented:
Won't doing `systemctl isolate node-image-pivot.target`have the effect of stopping the assisted/agent services that we need to avoid stopping?

jlebon commented:
> @jlebon That sounds like it might work. Where will the Kubelet be coming from? An OpenShift built image?

From the node image (i.e. for OCP, the `rhel-coreos` image in the release payload).

> Won't doing `systemctl isolate node-image-pivot.target`have the effect of stopping the assisted/agent services that we need to avoid stopping?

No. The system boots into `node-image-pivot.target` first. Any other service hooked into `multi-user.target` aren't started until after we've finished the live pivot.

cgwalters commented:
> The system boots into node-image-pivot.target first. 

Via a generator overriding `default.target`?

jlebon commented:
> > The system boots into node-image-pivot.target first.
> 
> Via a generator overriding `default.target`?

Yup, exactly. You can see that in https://github.com/openshift/installer/pull/8742.

zaneb commented:
The bootstrap ignition that's modified in https://github.com/openshift/installer/pull/8742 doesn't exist at boot time, so we'd need to also add this into the agent ISO and assisted Discovery ISO.
That's not impossible, but it'll be a challenge to maintain.
I guess if we are not able to pull the release image then nothing's going to work anyway, but doing it at boot time also limits our options for reporting any failure.

jlebon commented:
> The bootstrap ignition that's modified in [openshift/installer#8742](https://github.com/openshift/installer/pull/8742) doesn't exist at boot time, so we'd need to also add this into the agent ISO and assisted Discovery ISO. That's not impossible, but it'll be a challenge to maintain. I guess if we are not able to pull the release image then nothing's going to work anyway, but doing it at boot time also limits our options for reporting any failure.

Ahh right, tricky. Hmm, I think once the bootstrap bits are written out (by the `--once-from` MCD call), we could still just isolate back to `node-image-pivot.target` before continuing to bootstrapping. It should be possible to keep both the service and agent alive throughout. This will need more experimenting.

jlebon commented:
> so we'd need to also add this into the agent ISO and assisted Discovery ISO

A clarification on this: if we can split out those units somehow so they're accessible in the release payload and not baked into the installer, would whatever generates the ISOs be able to pull from the payload so we don't have to duplicate them across codebases? That'd make the more vanilla (non-AI/ABI) install flows more awkward though with a level of indirection.

zaneb commented:
> if we can split out those units somehow so they're accessible in the release payload and not baked into the installer, would whatever generates the ISOs be able to pull from the payload so we don't have to duplicate them across codebases?

We try to avoid having a hard dependency on the release payload in ABI, because the ISO may be generated in a different environment from where the cluster is going to run.

I think it could work to group all of the agent services under e.g. `agent.target` and in that target set `WantedBy: multi-user.target node-image-pivot.target`. That way I think we could isolate without stopping those services. Assuming it's safe to do the pivot while stuff is running.

jlebon commented:
> I think it could work to group all of the agent services under e.g. `agent.target` and in that target set `WantedBy: multi-user.target node-image-pivot.target`. That way I think we could isolate without stopping those services. Assuming it's safe to do the pivot while stuff is running.

Yeah, that's along the lines of what I was thinking as well. I like the `agent.target` idea.

I might reach out with questions for getting a test environment up to iterate on this.

yuqi-zhang commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+it fits perfectly in Prow's opinionated model of building the image and shoving it in the test
+release payload with which to launch test cluster.
+
+#### Re-organize MCO templates
```
Just wanted to add a :+1: from the MCO perspective

yuqi-zhang commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+- Maintain a single RHCOS stream per RHEL stream
+- Build the OpenShift node image as a layered image on top of RHCOS/SCOS
+- Build the OpenShift node image using Prow (in CI) and Konflux (in prod)
+- Introduce no additional reboots during cluster bootstrapping
```
To clarify, the initial update from bootimage -> current os image will still happen. This is to call out that we don't want a second reboot on top of this?

jlebon replied:
Right, exactly.

sdodson commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+on the RHEL side is that we would have only one stream of RHCOS per RHEL release,
+rather than one per OpenShift release. This greatly reduces the workload on the
+CoreOS team. Another benefit is easier integration in the CI processes of
+rhel-bootc and centos-bootc, as well as better shared documentation.
```
I don't that's accurate. The installation methods and composition of RHCOS today may yield certain differences in use but it's still RHEL and our position is still that it has the same hardware certification that RHEL carries.

cverna commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+benefits on the OpenShift side. Currently, RHCOS is built as a single layer that
+includes both RHEL and OCP content. This enhancement proposes splitting it into
+three layers. Going from bottom to top:
+1. the bootc layer; i.e. the base rhel-bootc image shared with image mode for RHEL (RHEL-versioned)
```
One thing that I have been thinking about, is that RHCOS often ships RHEL packages early (a few weeks before they are available in RHEL), in particular the kernel. These RHEL packages version will not be available in the rhel-bootc image and therefore would need to be replaced in the CoreOS layer, this happens often enough that I wonder if actually using the rhel-bootc image as base image for OCP is going to be useful.

@cgwalters, @sdodson and @jlebon what's your point of view for this?

jlebon replied:
This is why the sentence is left fuzzy; i.e. "shared with" instead of "deriving from". How exactly we'll share is still to be discussed. A key ingredient to success for me in this sharing is centralizing CI, and for that to be effective the NEVRAs in the RHCOS image ideally should be a superset of the NEVRAs in the rhel-bootc image that passed CI. And for that, we don't necessarily need to literally derive from.

That said, obviously the kernel is a pretty big one to deviate on, and we can manage that, but ideally we'd have a rhel-bootc stream that also consumes those early kernels since ISTM like it shouldn't interest only OCP, if only for testing purposes (though that widening of scope would need to be cleared by the kernel maintainers).

sdodson replied:
This could also just be a layer internal to the build process. I think the main goal is that we start from an image that RHCOS team doesn't produce, the RHEL team does.

jlebon replied:
@cverna, I'm going to mark this as resolved for now. It's not strictly in scope for this current enhancement, which is more concerned with splitting out the OCP bits first. But obviously something to keep in mind as we align with rhel-bootc.

jlebon commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,428 @@
+### Non-Goals
+
+- Change the cluster installation flow. It should remain the same whether IPI/UPI/AI/etc...
+- Introduce cluster administrator-visible changes. This change should be transparent to
+  admistrators. CoreOS layering instructions should keep working as is, but documentation
+  should ideally be reworked to leverage rhel-bootc docs more.
```
Added some thoughts about this in the "Risks and Mitigations" section. Let me know what you think there. Marking as resolved.

jlebon commented:
Updated and officially ready for review!
- Filled in the remaining sections
- Added details about versioning scheme
- Added details about solving the bootstrapping issue
- Added @yuqi-zhang and @jupierce as reviewers

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+
+#### Add more comprehensive CI
+
+Another important follow-up would to be to hook up better CI testing to openshift/os. There is
```
This is still possible even if ART continues to mirror rhcos into the CI imagestream. You would just disable ci-operator config "promotion" of the image. The openshift/os postsubmit job would then not push the CI image into the CI integration imagestream. 
ART would catch up with the changes later and push the latest updates to the CI integration imagestream.

jlebon replied:
Right, makes sense.

Re. promotion, let's continue that discussion in the first conversation above on that topic.

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+Once they're fully de-templatized, we can move them to openshift/os so that they
+are part of the node image proper.
+
+This would considerable shrink the amount of templates that the MCO would
```
```suggestion
This would considerably shrink the amount of templates that the MCO would
```

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+OCP-versioned directories (e.g. `4.16.3/`). We don't want customers to have to
+know which version of RHCOS bootimages to download for their target OCP version.
+So to avoid confusing customers, we should still have OCP-versioned directories
+but it would be wasteful to not symlink the bootimages on the server side to
```
Unfortunately, mirror.openshift.com doesn't support symlinks.

jlebon replied:
Ouch, that's unfortunate. Or ideally just any kind of content-addressed storage. But anyway, softened the wording there.

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+Regarding other OCP components, the goal is to catch these dependencies during
+CI testing of the installer PR that switches over to the new bootimages. We
+need to raise awareness of the change to the whole OCP org and call for extra
+testing. This needs to be a coordinated event with ART, DPTP, and TRT. Of
```
```suggestion
testing. This needs to be a coordinated event with ART, CRT, and TRT. Of
```

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+  the node layer is RHEL-versioned and possibly shared across multiple OCP
+  releases, any fast-tracks would need to happen by doing an RPM upgrade within
+  the node layer. This is more cumbersome than just tagging packages into the
+  the ART plashets as currently done, but makes us more thorough in our known
```
Can you unpack "cumbersome"?
This seems vital to deliver a hotfix. I can understand it not being ideal from an image size perspective, but still possible, right?

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+    Brew and those builds then being added to the ART plashets. To avoid having
+    to source the (OCP-versioned) ART plashets during the CoreOS builds, we
+    could instead create new repos for these packages that are clearly defined
+    to be RHEL-versioned. The CoreOS build could then add that repo alongside
```
I think ART would need to dnf upgrade the kernel in the node layer from their plashet. If the coreos layer is up-to-date, it is a no-op. Otherwise, we lose some of the benefit of the layer decoupling.

jlebon replied:
Hmm, when would it be not up to date? The ART plashet sources it from the same place the RHCOS image would source it from. It's certainly possible to replace the kernel in the node image build, but IMO it's much more appropriate to respin the RHCOS base image for it.

jupierce replied:
> The ART plashet sources it from the same place the RHCOS image would source it from

Seems like I misinterpreted "we could instead create new repos for these packages that are clearly defined to be RHEL-versioned" which sounded like you would not be consuming ART generated repos. Reading it now, I think you mean that the source would still be created by ART, but RHEL-version specific (e.g. plashets/el9.2 instead of plashets/4.19). 

This would mean we have no ability to have different kernel versions between different Z's using the same RHEL (tagging in a kernel fix for 4.x could impact RHCOS for 4.y). @joepvd concerns here?

In any case, ART needs to ability to create hotfixes including kernel changes. This would need to be layered on top of the RHCOS image unless RHCOS will allow us to trigger custom builds as they do today.

jlebon replied:
> Reading it now, I think you mean that the source would still be created by ART, but RHEL-version specific (e.g. plashets/el9.2 instead of plashets/4.19).

Yes, exactly. I'll clarify this.

> This would mean we have no ability to have different kernel versions between different Z's using the same RHEL (tagging in a kernel fix for 4.x could impact RHCOS for 4.y).

To me, that's a feature, not a bug. :) Differing from RHEL should be the exception not the rule, so having that tracked in the layered image instead only when needed seems appropriate.

> In any case, ART needs to ability to create hotfixes including kernel changes. This would need to be layered on top of the RHCOS image unless RHCOS will allow us to trigger custom builds as they do today.

Ahh yes, thanks for bringing up hotfixes. I think in general the pipecfg hotfix path can be considered deprecated by layering. So instead, you'd do a custom layered build where you can e.g. apply hotfixes for whatever packages you want there. The exception is if it's a hotfix required for firstboot though I suspect that's much less common. (But even then, long-term us or users can transform the layered image into bootimages.) I'll mention this.

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+- the OpenShift installer now uses the new bootimages
+- the OpenShift node image can be built in Prow/Konflux
+- the new artifacts are used as part of the CI streams and all tests pass
+
```
- Release controller updates for diff view. cc @bradmwilliams

jlebon replied:
Thanks, added!

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+
+- Work with DPTP(?) to update the OCP release page to better handle the new layered approach.
+    - It currently links to diffs in the RHCOS release browser. But now the release browser
+      will only contain the RHEL diff, not the OCP diff in the node layer. We should close this
```
This would also break some ART logic. They query the release browser to determine the version of RHEL and OCP package versions in a given version to ensure that the images are up-to-date.

jlebon commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+container image, just like most other OpenShift components are built. This should
+allow simplifying CI and ART tooling related to RHCOS, much of which is bespoke.
+
+For example, ART would no longer have to sync RHCOS images to CI; CI could
```
Thanks, that's a detail I didn't know but it makes sense. So overall images in the CI imagestreams may come from either CI _or_ ART nightlies depending on how fresh the repo is?

I'm trying to verify my understanding of this by comparing the image digests in e.g. registry.ci.openshift.org/ocp/release:4.17.0-0.ci-2024-09-24-105044 vs registry.ci.openshift.org/ocp/release:4.17.0-0.nightly-2024-09-23-211417; I would've expected some of them to be the same there. The CI payloads are manifest listed but not the nightly ART one, but even comparing against the amd64 digests in the manifest lists against the nightly ones, I'm not getting any component images that match other than the RHCOS ones.

Hmm, and based on [this other comment](https://github.com/openshift/enhancements/pull/1637/files#r1755881465), are you suggesting not having CI builds promote at all and keep relying on just ART or both? I'd like the CoreOS images to be at parity with the other component images. E.g. merges to openshift/os would ideally trigger a CI build and promotion.

jupierce replied:
> I'm not getting any component images that match other than the RHCOS ones.

ART pushes RHCOS and _base_ images to the _integration_ imagestream. The latter serve as base images for other component builds that wind up in the release payload (e.g. https://github.com/openshift/oc/blob/3692450b96d57ae3870e5f693e833a0701d0e2b0/images/cli/Dockerfile.rhel#L6 ) . The base images themselves do not show up in the payload.

> not having CI builds promote at all and keep relying on just ART or both?

We could:
1. Leave things as they are today (ART pushes RHCOS imagestreams to the CI integration imagestream).
2. Have only CI promote rhcos into the CI integration imagestream.
3. Have both push.

The downside of #1 is that is not fast. It may take hours for an RHCOS merge to reach the CI integration imagestream by way of ART.

The downside of #2 is that the production integration imagestream can diverge, for significant periods of time, from the CI integration imagestream. CI builds & promotion generally only happen on merge in the component's repository. A merge in openshift/hyperkube with important updates to the kubelet would not cause #2 to rebuild RHCOS. This can lead to difficult to diagnose disparities (e.g. nightlies passing tests but CI payloads failing).

The downside of #3 is that it is obscure and provenance of the RHCOS image in the CI integration stream can flip back and forth between ART and CI builds. ART should also put in a check to ensure that it doesn't update the imagestream with an image that is older than one CI has already promoted.

jlebon replied:
> ART pushes RHCOS and base images to the integration imagestream. The latter serve as base images for other component builds that wind up in the release payload

Gotcha, I follow now. ART does pushes to e.g. registry.ci.openshift.org/ocp/4.18:base-rhel9. This then triggers a rebuild _in the CI cluster_ of all the images based on that base image.

> We could:
>  1. Leave things as they are today (ART pushes RHCOS imagestreams to the CI integration imagestream).
>  2. Have only CI promote rhcos into the CI integration imagestream.
>  3. Have both push.

Before digging into that, let me clarify again the images in play here:
- there's the CoreOS base image containing the pure RHEL/CentOS Stream content (I've been calling this the "rhel-coreos-base" image)
- there's the OCP node image which builds on top of the rhel-coreos-base image and adds all the OCP-specific bits (which confusingly will still be called "rhel-coreos" in the release payload but ideally we'd rename to e.g. "node")

The rhel-coreos-base image will keep being built in the internal CoreOS pipeline (currently Jenkins, in the future Konflux).

What I'm trying to see is if it makes sense to have the OCP node image be setup in a similar way to the other OCP components. So e.g. taking openshift/oc as an example, which is built on top of the base-rhel9 base image. It has both a CI build which gets promoted to the CI imagestream and a prod ART build which gets pushed to the nightly imagestream (right?).

So then, two approaches I see for handling RHCOS are:
1. Only have ART build the node image. It then syncs it to the CI imagestream.
2. Have an ART build of the node image _and_ a CI build of the node image. The CI build is triggered by changes in openshift/os, but also by changes to its base image (rhel-coreos-base), which ART or the CoreOS pipeline itself would push to the CI imagestream.

(1) is closer I think to what we're currently doing. (2) is closer to how other OCP components are set up and is what I'm currently proposing.

As you correctly noted, one key aspect of the node image is that it consumes content from other OCP components as RPMs built by ART. As mentioned in Phase 3, ART will have to adapt its logic to detect if OCP RPMs in the node image are stale and we need to rebuild. In (2), it could be the one that also retriggers the CI build at the same time it triggers its own build to ensure CI doesn't go stale but without muddling provenance.

jlebon replied:
Chatted with Justin about this. There was agreement to keep things simple to start and just have ART also push to the CI imagestream.

jlebon commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+
+* As an OpenShift customer, I want to be able to follow (most of) the image mode for RHEL documentation and use (most of) the same tooling surrounding it when customizing RHCOS, so that I have a consistent experience across image mode for RHEL and RHCOS and a clear understanding of how they relate to each other.
+
+* As an ART engineer, I want to be able to build the OpenShift node image in a similar way to how the other OpenShift components are built, so that I can simplify my tooling and pipelines.
```
Right, yeah. Though note the bootc images are base image builds whereas the node image here would be a layered image builds and do not require additional permissions. Those should build fine in e.g. OSBS today AIUI, though yes I think ideally we go straight to Konflux as one of the early adopters.

So then, the user story here is that that node image build isn't too special and e.g. the Konflux pipeline definitions can be reused between OCP components pretty well.

jlebon commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,458 @@
+which will build the OpenShift node image by layering the OpenShift components on top: RPM
+packages, config files, etc...
+
+After some testing, the Prow job will be configured to push to the `rhel-coreos` tag in the CI
```
Agreed. Will update this section to make it match what we want after the conversation above about this is resolved.

LorbusChris commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,467 @@
+
+## Proposal
+
+The overall proposal consists of two parts. In the first part, we will focus on moving
+the OCP-specific packages out of RHCOS and into its own layer (the node layer). In the second
+part, we'll change RHCOS itself to become a layered image (the CoreOS layer) built on top
+of the rhel-bootc work (the bootc layer).
```
We want the node layer to also contain static config files that currently live in MCO as yaml.

@cgwalters

LorbusChris replied:
Also stated in https://github.com/openshift/enhancements/pull/1637/files#diff-13a1c8d3015c20328d82dd45e44f78e6ddc9d98d6b1b0b5077345b5fe82b0055R286

jupierce commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,467 @@
+
+- Work with CRT to update the OCP release page to better handle the new layered approach.
+    - It currently links to diffs in the RHCOS release browser. But now the release browser
+      will only contain the RHEL diff, not the OCP diff in the node layer. We should close this
```
The release browser would also report the wrong thing if, for example, ART layers on a new kernel. The release controller does not presently have a method of reporting RPM versions without the release browser, so we will lose visibility into the hyperkube version installed in the `node` image from the release controller.

cgwalters replied:
"release browser" should go away as a concept: what's in the container image can be found by introspecting the container image.

We must handle the general case of "higher layer overrides/removes content from lower layer" across all containers - the OS is not special.

jupierce replied:
Agreed. But the release controller is stateless and can't introspect all images in a reasonable period of time for a payload details page to render. So I suspect there is significant work to explain here. `oc adm release new` would have the time and access necessary to store the RPM information and it could be new metadata output in `oc adm release info -o json` .

cgwalters replied:
AIUI today the diff is dynamic, but I can't think of a problem with us maintaining a long term cache of the extracted rpmdb for each image. I think quay.io may actually do something like this already? The [security scan](https://quay.io/repository/fedora/fedora-bootc/manifest/sha256:7bc5ec286a804d89de2b739725fa82ad6dbb8a17668197d612760ec7ab90ffe3?tab=vulnerabilities) is already has is basically a rendered derivation of that.

Then given that cache, we can compute the diff between two container images easily, which is what the release browser does today.

jupierce replied:
> diff is dynamic

`oc adm release info --changelog` is used. It uses information baked into the release payload to find the commits associated with each of the payload components. `oc` then uses git repo clones to quickly compute the diff based on the commit history.

This changelog output from `oc` is in markdown and the release controller just translates it into HTML. The release controller doesn't presently have any means to interject new elements, like an RPM diff, in the contiguous markdown document without brittle heuristics. 

changelog also doesn't deal with heterogeneous payloads today as commits must agree between all arches of a component. This is not the case for different architectures & RPMs.

So a list of options:
1. Release controller introspects images. Significant change to behavior today. New cache (quay does not have information on registry.ci.openshift.org images -- required for non-OS).
2. `oc` does the introspection. `oc ` includes the RPM data in markdown. quay is not an option because payload component images are not necessarily on quay.io (and it seems unreasonable to require users to have a quay token). New views / choreography with release controller unless RPMs are included inline within the markdown (could be voluminous / difficult to read). 
3. Something else?

#2 makes sense to me. In any case, I feel there are weeks of work here for one or more teams we shouldn't discount if the release browser disappears.

jlebon replied:
I'll distill this conversation in the enhancement, but ISTM we don't need to have it fully fleshed out before getting this merged. But yes, definitely a gap to address before rolling out to prod.

jlebon replied:
Chatted with Justin about this a bit more and did some tests. It's very expensive to extract the rpmdb. First, currently it seems like `oc image extract` chokes on hardlinks so we'd have to fix that. I think one hack would be to always end the node image build with e.g. `RUN rpm -qa > /usr/lib/sysimage/rpm-base.txt` or something, and then `oc image extract` could much more cheaply pull down just the final layer. But yuck...

Anyway, I'm thinking the simplest is probably to just keep using the release browser. Have a service that just caches the rpmdb from built node images and pushes it out to S3 in a way that the release browser already knows how to consume.

sdodson commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,467 @@
+Ideally the mirror server could symlink/somehow dedupe the bootimages shared
+across OCP versions but this is of course not a requirement.
+
+As such, the versioning change should be transparent to end-users, but on the
+off chance that a customer is relying on the current semantics, we should still
+announce it in the GA release notes.
```
And we'd start to see installer PRs moving 4.19-4.21 all to the same boot images?

LorbusChris replied:
Tangentially related to this is also moving static templates from the MCO to the coreos overlay: https://issues.redhat.com/browse/MCO-1429

wking replied:
> Always have the installer fetch it from the configmap in the target release payload.

At the moment (or at least as of my last installer sync, I might be stale), there's no requirement for the installer to be able to access the release image it's launching.  E.g. an installer could be talking to cloud-provider APIs to create infra, and to the Kube API via a proxied pipe, but it does not need access to the registry holding the release image.  You could maybe dance around that by grabbing boot image data and stuffing it into the installer binaries at build time, although I'm not sure how complicated that would be at the build-tooling layer.  Or you could also teach `oc adm release extract ...` to inject it at extraction time (as [it already does with some other release properties][1]).

Or maybe the installer functioning without access to the release-image registry isn't that big a deal.  I'd taken a pass at this kind of access way back in openshift/installer#1286, and it seems to have failed more on "meh, not worth the effort" than on "this adds registry-access dependencies we don't want to carry".

[1]: https://github.com/openshift/oc/blob/8e9a2be04f902d58b3be21b92ed9016854640cd6/pkg/cli/admin/release/extract_tools.go#L57-L59

andfasano replied:
> there's no requirement for the installer to be able to access the release image it's launching
Note that ABI currently does require to access the release image during the installation for various bits. In particular, for the disconnected installations, the live ISO image is [extracted](https://github.com/openshift/installer/blob/19c87c40a42acd92323c32b62af8de605498e5bb/pkg/asset/agent/image/baseiso.go#L216) from the `machine-os-images` release image

jlebon commented:
Waiting on final review.
/remove-lifecycle stale

mrunalp commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,514 @@
+
+### User Stories
+
+* As an OpenShift engineer, I want the openshift/os repo to work similarly to other OpenShift repos (with a `Containerfile`), so that CI can easily build and test a PR in a cluster and humans too can more easily build and test PRs or e.g. use Cluster Bot.
```
nit: humans, too, can ..

mrunalp commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,514 @@
+
+### Non-Goals
+
+- Change the cluster installation flow. It should remain the same whether IPI/UPI/AI/ABI/etc...
```
nit etc.

mrunalp commented on enhancements/rhcos/split-rhcos-into-layers.md:
```diff
@@ -0,0 +1,514 @@
+      "Open Questions" below).
+- Work with ART to adjust how they determine if the RHCOS images are up to date.
+  Currently, they expect both RHEL and OCP content in those images, but that content will
+  be split across separate images now.
```
wouldn't the node image built on top of the base image have all the contents?

jlebon replied:
It would. I'm not sure on the details there of how the ART machinery works, but my understanding is that RHCOS builds currently get triggered as part of a larger pipeline that also builds e.g. the kubelet RPM. But with this change, there's no point in rebuilding RHCOS if just OCP packages changed. Instead, a new layered build should happen.

mrunalp commented:
Left some nits but looking good overall.

jlebon commented:
Updated for comments!

mrunalp reviewed (approved):
/lgtm
